### PR TITLE
AlzViT (Visual Transformer for Alzheimer's Detection) - Zachary Walls - 46966775

### PR DESCRIPTION
**Pull request by zacwallsie**
## Overview

This pull request implements a Visual Transformer (ViT) design reflective of the paper it was first introduced in "An Image Is Worth 16x16 Words" by Dosovitskiy et al. with some slight alterations to improve performance. The model was trained and tested on the ADNI dataset which contains a number of MRI brain images split into 2D slices being labeled as either cognitive normal (NC) or Alzheimer's disease (AD).

The model achieved an impressive 83.06% accuracy on the testing dataset supplied in the ADNI files utilising a model trained for 50 epochs on the training dataset, more details of which can be found in the `README`.

## Project Structure

The project is organised into several files and folders:

- `modules.py`: Contains the implementation of the ViT model components, including patches, patch encoding, and the main classifier.
- `dataset.py`: Contains data loading and preprocessing functions to prepare the image dataset for training.
- `train.py`: Trains the ViT model on the provided dataset and saves the trained model.
- `predict.py`: Uses the trained model to make predictions on single images and visualize the results.
- `requirements.txt`: Requirements file including the dependencies and libraries used throughout the scripts.
- `images/`: Contains the images within the `README`.
- `README`: Contains details of the results and how to train and test the model.

Overall the ViT Model yielded promising results however could be improved upon with introductions of other models such as CNNs to improve the performance of the model architecture as noted on in the `README`.

### CODE REVIEW

**Comment by LinfengLiu98** [comment]
**This is an initial inspection, no action is required at this point**

Difficulty: Hard

Readme: 
- Excellent and detailed


Commit messages: Excellent

Code:
- Uses Vision transformer
- Well commented
- Good design

Functionality/Performance:
- Excellent work with 83.06% acc test accuracy.

Best work I have ever seen so far! Good job :)

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe good
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
